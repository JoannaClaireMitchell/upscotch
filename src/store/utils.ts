import * as Types from "./types";

export function createNewList(listId: number) {
  return {
    description: `Tasks in list ${listId}`,
    listId,
    name: `List ${listId} Name`,
    order: listId,
  };
}

export function generateListId(lists: Types.List[]) {
  let number = lists[lists.length - 1].listId + 1;
  while (lists.some((list) => list.listId === number)) {
    number++;
  }
  return number;
}

export function splitToLists(
  data: {[listId: number]: Types.Task},
  lists: number
): Types.TaskList {
  const columns = Object.values(data).reduce((acc, item) => {
    if (!acc[item.listId]) {
      acc[item.listId] = [item];
    } else {
      acc[item.listId] = [...acc[item.listId], item];
    }
    return acc;
  }, {} as Types.TaskList);

  if (Object.keys(columns).length < lists) {
    for (let index = 1; index <= lists; index++) {
      if (!columns[index]) {
        columns[index] = [];
      }
    }
  }

  return columns;
}
