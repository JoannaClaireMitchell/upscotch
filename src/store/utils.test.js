import {splitToLists, generateListId} from "./utils";

const items = {
  1: {
    id: 1,
    order: 1,
    listId: 1,
    name: "first",
    description: "Geese  can fly",
  },
  2: {
    id: 2,
    order: 2,
    listId: 2,
    name: "second",
    description: "Nom Nom",
  },
  3: {
    id: 3,
    order: 3,
    listId: 1,
    name: "third",
    description: "Sugar cubes",
  },
};

describe("splitToLists", () => {
  it("returns split lists", () => {
    const value = splitToLists(items);
    expect(value).toMatchSnapshot();
  });
  it("can move first item into another list", () => {
    const values = splitToLists({...items, 1: {...items["1"], listId: 2}});
    expect(values).toMatchSnapshot();
  });
  it("can allow all items in second list", () => {
    const values = splitToLists({
      ...items,
      1: {...items["1"], listId: 2},
      3: {...items["3"], listId: 2},
    });
    expect(values).toMatchSnapshot();
  });
  it("allows removal of a middle list", () => {
    const values = splitToLists({
      1: {...items["1"], listId: 1},
      2: {...items["2"], listId: 3},
      3: {...items["3"], listId: 3},
    });

    expect(values).toMatchSnapshot();
  });
});

describe("generateListId", () => {
  it("should return the next unused number", () => {
    const id = generateListId([
      {listId: 1},
      {listId: 3},
      {listId: 5},
      {listId: 7},
      {listId: 8},
    ]);
    expect(id).toEqual(9);
  });
});
