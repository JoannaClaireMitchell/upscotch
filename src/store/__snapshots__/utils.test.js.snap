// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`splitToLists > allows removal of a middle list 1`] = `
{
  "1": [
    {
      "description": "Geese  can fly",
      "id": 1,
      "listId": 1,
      "name": "first",
      "order": 1,
    },
  ],
  "3": [
    {
      "description": "Nom Nom",
      "id": 2,
      "listId": 3,
      "name": "second",
      "order": 2,
    },
    {
      "description": "Sugar cubes",
      "id": 3,
      "listId": 3,
      "name": "third",
      "order": 3,
    },
  ],
}
`;

exports[`splitToLists > can allow all items in second list 1`] = `
{
  "2": [
    {
      "description": "Geese  can fly",
      "id": 1,
      "listId": 2,
      "name": "first",
      "order": 1,
    },
    {
      "description": "Nom Nom",
      "id": 2,
      "listId": 2,
      "name": "second",
      "order": 2,
    },
    {
      "description": "Sugar cubes",
      "id": 3,
      "listId": 2,
      "name": "third",
      "order": 3,
    },
  ],
}
`;

exports[`splitToLists > can move first item into another list 1`] = `
{
  "1": [
    {
      "description": "Sugar cubes",
      "id": 3,
      "listId": 1,
      "name": "third",
      "order": 3,
    },
  ],
  "2": [
    {
      "description": "Geese  can fly",
      "id": 1,
      "listId": 2,
      "name": "first",
      "order": 1,
    },
    {
      "description": "Nom Nom",
      "id": 2,
      "listId": 2,
      "name": "second",
      "order": 2,
    },
  ],
}
`;

exports[`splitToLists > returns split lists 1`] = `
{
  "1": [
    {
      "description": "Geese  can fly",
      "id": 1,
      "listId": 1,
      "name": "first",
      "order": 1,
    },
    {
      "description": "Sugar cubes",
      "id": 3,
      "listId": 1,
      "name": "third",
      "order": 3,
    },
  ],
  "2": [
    {
      "description": "Nom Nom",
      "id": 2,
      "listId": 2,
      "name": "second",
      "order": 2,
    },
  ],
}
`;
